docs/madr/decisions/056-meshexternalservice-and-zoneegress.md

```diff
@@ -0,0 +1,407 @@
+
+It naturally fits our targetRef mechanism.
+It does not make sense to apply inbound policies on the sidecar since there is no inbound for it.
+Egress plays a role of a "sidecar" that is inside the mesh, and we can target it's inbound.
+Configuring outbound policies on the sidecar avoids the problem with different clients influencing each other (see [MeshCircuitBreaker](#meshcircuitbreaker)).
+
+##### How to configure it
+
+###### From ONLY policy (policies that do not have to)
+
+Policies:
+- MeshTrafficPermission
+- MeshRateLimit
+- MeshGlobalRateLimit
+- MeshFaultInjection
+
+```yaml
+kind: MeshTrafficPermission
+spec:
+  targetRef:
+    kind: MeshExternalService
+    name: httpbin
+    proxyType: egress
+  from:
+    - targetRef:
+        kind: MeshService
+        name: backend
+      default:
+        action: Allow
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Deny
+```
+
+###### `To` policy on sidecar
+
+Policies:
+- MeshRetry
+- MeshHTTPRoute
+- MeshTCPRoute
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshRetry
+metadata:
+  name: backend
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+        kind: MeshExternalService
+        name: backend
+      default:
+        http:
+          numRetries: 10
+```
+
+###### `To` policy on egress
+
+The following policy we're going to configure on Egerss: 
+- `MeshCircuitBreaker` - we don't want to cut out whole egress from traffic only faulty instance of MeshExternalService
+- `MeshLoadBalancingStrategy` - (we only care about `default.loadBalancer` portion) we don't want to change the loadbalancing to Egress but to the underlying MeshExternalServices
+- `MeshHealthCheck` - we don't want to be health checking Egress but the underlying MeshExternalServices
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    proxyType: egress
+  to:
+    - targetRef:
+        kind: MeshExternalService
+        name: backend
+      default:
+        outlierDetection:
+          detectors:
+            totalFailures:
+              consecutive: 10
+```
+
+###### `To` and `From` policies
+
+Policies:
+- MeshAccessLog
+- MeshTimeout
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshAccessLog
+spec:
+  targetRef:
+    kind: Mesh
+    proxyType: egress
+  to:
+  - targetRef:
+      kind: MeshExternalService
+      name: backend
+    default: ...
+```
+
+##### Empty `proxyType`
+
+Currently, egress is not selected by `kind: Mesh` targetRef, but missing `proxyType` means both `gateway` and `sidecar`.
+We don't want to assume empty `proxyType` selecting `egress` because we want greater control over what is configured and to make sure user is aware that egress is affected.
+To avoid breaking changes now and in the future we can extend this more easily.
+
+We will require `proxyType: egress` in validation:
+- on top level `MeshExternalService`
+- on to egress policies
```

Comment:
Not sure I understand this, what happens if I apply:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshAccessLog
spec:
  targetRef:
    kind: Mesh
  to:
  - targetRef:
      kind: MeshExternalService
      name: backend
    default: ...
```

Reply:
MeshAccessLog is `to&from` policy, here we're talking about `MeshHealthCheck`, `MeshLoadBalancingStrategy`, `MeshCircuitBreaker`. Right @lukidzi ?

Reply:
I think Ilya's question is where config is applied with the policy he proposed:
1) It configures MAL only on sidecar
2) It configures MAL on sidecar and egress

Which one is it?

Reply:
I would say both. A better example is what @lukidzi mentioned with timeouts. If you apply a timeout you probably don't want different timeout on the sidecar vs the egress right?

Reply:
But then it contradicts the statement, no?
> We don't want to assume empty `proxyType` selecting `egress` because we want greater control over what is configured and to make sure user is aware that egress is affected.

For policies like MeshRetry we probably don't want to have both `sidecar` and `egress` by default.

Reply:
You're right :| I think we need to always require `proxyType` when `MeshExternalService` is referenced and document the Timeout case.